### ISSUE TITLE
The prediction result of Temporal Fusion Transformer is a straight line

### ISSUE THREAD

**Issue by 2448778655** [jdb78/pytorch-forecasting#344]
- PyTorch-Forecasting version: 0.8.3
- PyTorch version: 1.7.1
- Python version: 3.9
- Operating System: win10

### Expected behavior

I execute the **Temporal Fusion Transformer** for power bus load forecasting. The resolution of the data is 15 minutes, which means that there are 96 data points in a day. I want to enter the data from the previous day to predict the 96 data values of the day.

### Actual behavior

But the predicted curve is a straight line. The predicted result is 96 nearly the same value.
![image](https://user-images.githubusercontent.com/40781567/108014869-dbf32280-7049-11eb-87d6-0efcf03074eb.png)

**Comment by jdb78** [comment]
Do you have a colab notebook available?

**Comment by jdb78** [comment]
BTW: You need to add your target variable to the `time_varying_unknown_reals`

**Comment by 2448778655** [comment]
> BTW: You need to add your target variable to the `time_varying_unknown_reals`

I followed your suggestion, but it did not solve the problem.

Here is my colab:

https://colab.research.google.com/drive/1sJKz90011SJXEPU8iIreZ06aAQAzz2v2?usp=sharing

**Comment by 2448778655** [comment]
> BTW: You need to add your target variable to the `time_varying_unknown_reals`

If I understand correctly this should be a multi-horizon prediction task. How should I set the value of horizon?

**Comment by jdb78** [comment]
The horizon is the `max_prediction_length`. Maybe also try to add the `time_idx` to known reals and try to overfit using the GroupNormalizer for target normalization. Generally speaking your problem might not be very well suited to deep learning given that there is only one time series with around 3000 samples.

I could have a deeper look if you can make the data available to public (currently, the colab notebook requires the data to be on one's private GDrive).

**Comment by 2448778655** [comment]
I have sent you the link to the data file via the email address you left on LinkedIn.

BTW: Could you please explain in detail the meaning of these four variables: min_encoder_length, max_encoder_length, min_prediction_length, max_prediction_length?

**Comment by zvit3k** [comment]
Hi guys,

max_encoder_length - your maximum historical data to be taken into consideration for training
min_encoder_length - minimum length to be taken into consideration (for which prediction will be provided), when set to 1, you accept the prediction with no history
max_prediction_length - like Jan mentioned above, horizon you would like to be taken into prediction 
min_prediction_length - (probably) like explanation for min_encoder_length, but with this one we should ask @jdb78 

What is more, I would like to refresh topic because I am facing similar problem with flat line:

PyTorch forcasting: 0.8.3
PyTorch lightning: 1.1.8
Python: 3.7
Debian Buster 10

<img width="491" alt="chart" src="https://user-images.githubusercontent.com/16278338/109561059-f531af80-7adc-11eb-893a-876f783fc2ca.PNG">

Others examples ( when plotting ) looks similar. I cannot share the data itself but I will try provide as many details I can:

- Time_idx count : 5k (too few examples?)
- Time_idx max : 399 (days)
- Data extended with additional day of week column 
- No special days included 
- Values in columns are in range from -1 to 1.

I am changing the training parameters, but I think this is not the case. I probably missed something important.
Thanks for support!
Witold

**Comment by 2448778655** [comment]
> Hi guys,
> 
> max_encoder_length - your maximum historical data to be taken into consideration for training
> min_encoder_length - minimum length to be taken into consideration (for which prediction will be provided), when set to 1, you accept the prediction with no history
> max_prediction_length - like Jan mentioned above, horizon you would like to be taken into prediction
> min_prediction_length - (probably) like explanation for min_encoder_length, but with this one we should ask @jdb78
> 
> What is more, I would like to refresh topic because I am facing similar problem with flat line:
> 
> PyTorch forcasting: 0.8.3
> PyTorch lightning: 1.1.8
> Python: 3.7
> Debian Buster 10
> 
> <img alt="chart" width="491" src="https://user-images.githubusercontent.com/16278338/109561059-f531af80-7adc-11eb-893a-876f783fc2ca.PNG">
> 
> Others examples ( when plotting ) looks similar. I cannot share the data itself but I will try provide as many details I can:
> 
> * Time_idx count : 5k (too few examples?)
> * Time_idx max : 399 (days)
> * Data extended with additional day of week column
> * No special days included
> * Values in columns are in range from -1 to 1.
> 
> I am changing the training parameters, but I think this is not the case. I probably missed something important.
> Thanks for support!
> Witold

Thank you for your sharing. I have tried to modify the parameters, but the problem has not been solved. I don't know where I went wrong, if you solve the problem I hope you can share your solution ~

**Comment by dempseyryan** [comment]
What are your current hyper parameters? I had this issue a while back but after revising a bunch the model is now doing well. I found hidden size to be one of the more important parameters, as suggested in the tutorial. It could be that your model is just not currently complex enough to predict your data's patterns. So I would say try cranking the hidden size and continuous hidden size as much as your set-up can handle to attempt to overfit your data. Though as @jdb78 mentions it might just be difficult for the model without more training examples.

**Comment by zvit3k** [comment]
Hi

Thank you for your thoughts. I have tried experimenting with hyperparameters but with no results. This is the last experiment:
max_epochs=20, 
limit_train_batches=25,
gradient_clip_val=0.1,
learning_rate=0.6 (trainer found learning rate at 1.23, but it gave the same output->flatlines)
hidden_size=128,
attention_head_size=2,
dropout=0.3,
hidden_continuous_size=64,
output_size=7,
reduce_on_plateau_patience=4

Are there any suggestions when it comes to number of series and samples per series to make a a model working without any problems ? Or maybe there are some rules when it comes to setting the encoder length / prediction etc ?

Thanks
Witold

**Comment by dempseyryan** [comment]
Happy to help. I just started learning about machine learning in January so take my thoughts with a grain of salt. Perhaps increase the number of batches per epoch as well as training for more epochs. Is your validation loss still on a downward trend when the training has been stopping? If so, then you should try training longer and see what happens. Also is that learning rate suggested from the finder? It's pretty high -- the network might be continually overshooting its path to minimum loss.

As for length suggestions, as far as I'm aware it's very data-dependent but typically the longer encoders and shorter prediction are easier on the model. As for number of series and samples I'm not sure an exact number but the more the better.

Again I'm pretty new to this stuff so my suggestions are based on very little experience.

**Comment by zvit3k** [comment]
The learning rate suggested from the finder is 1.23 - I was also shocked, because normally those values are quite low. I removed the limit -> limit_train_batches and still nothing. Early stopping is on, so after 19 epochs it stops automatically. 

Witold

**Comment by dempseyryan** [comment]
Interesting. Have you compared validation losses each epoch? How is your early stopping defined? I'm thinking it might be not learning much and it stops early not because the validation has gotten really good but because it just isn't learning (which results in the validation not improving, which results in early stopping)? The only thing I can think of with the learning rate is that with your amount of data it's just too noisy but honestly I'm not sure.

**Comment by 2017wxyzwxyz** [comment]
@2448778655，Do you have the data of this project ’stallion.ipynb‘  ?

**Comment by cowwoc** [comment]
@2448778655 I am having the same problem. How did you end up solving this?

**Comment by cmajorsolo** [comment]
@2448778655 I am having the same issue. I wonder if it is related to the seasonality and trend in the dataset? Could you shed light on how you resolved the problem?
